float32[] hi
float32[] lo